pkg/storegateway/indexheader/binary_reader.go

```diff
@@ -417,6 +418,36 @@ type postingValueOffsets struct {
 	lastValOffset int64
 }
 
+func (e *postingValueOffsets) prefixOffset(prefix string) (int, bool) {
+	// Find the first offset that is greater or equal to the value.
+	offsetIdx := sort.Search(len(e.offsets), func(i int) bool {
+		return prefix <= e.offsets[i].value
+	})
+
+	// We always include the last value in the offsets,
+	// and given that value is always less or equal than the value,
+	// we can conclude that there are no values with this value.
+	if offsetIdx == len(e.offsets) {
+		return 0, false
+	}
+
+	// Prefix is lower than the first value in the offsets, and that first value doesn't have this value.
+	// Next values won't have the value, so we can return early.
+	if offsetIdx == 0 && prefix < e.offsets[0].value && !strings.HasPrefix(e.offsets[0].value, prefix) {
+		return 0, false
+	}
+
+	// If the value is not equal to the value, this value might have the value.
```

Comment:
```suggestion
	// If the value is not equal to the prefix, this offset might have the prefix.
```

Is this what you meant?

Reply:
Oh dear, fixed all the comments in https://github.com/grafana/mimir/pull/4055/commits/13a551fd69812e78ee71f3cf1797eb64665f5016

I refactored this method renaming `prefix` to `value` because I wanted to reuse it in `PostingsOffset`, then I saw the small difference (the prefix check) and renamed back, and that's where this mess comes from.